v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/dialectadapter/mysql/MysqlDialectAdapter.java

```diff
@@ -110,6 +158,58 @@ public ImmutableMap<String, ImmutableMap<String, SourceColumnType>> discoverTabl
     return tablesBuilder.build();
   }
 
+  /**
+   * Discover the indexes of tables to migrate.
+   *
+   * @param dataSource Provider for JDBC connection.
+   * @param sourceSchemaReference Source database name and (optionally namespace)
+   * @param tables Tables to migrate.
+   * @return The discovered indexes.
+   * @throws SchemaDiscoveryException - Fatal exception during Schema Discovery.
+   * @throws RetriableSchemaDiscoveryException - Retriable exception during Schema Discovery.
+   */
+  @Override
+  public ImmutableMap<String, ImmutableList<SourceColumnIndexInfo>> discoverTableIndexes(
+      DataSource dataSource,
+      SourceSchemaReference sourceSchemaReference,
+      ImmutableList<String> tables)
+      throws SchemaDiscoveryException, RetriableSchemaDiscoveryException {
+    String discoveryQuery = getIndexDiscoveryQuery(sourceSchemaReference);
```

Comment:
Why are we discovering indexes and not primary keys directly?

Reply:
In the longe-run we will support any unique index with good cardinality. Currently we are limiting to primary keys to limit the scope of support and testing for M1.